src/Client.php

```diff
@@ -79,6 +91,170 @@ public function __construct(Transport $transport, LoggerInterface $logger = null
+        }
+
+        $this->logger->debug('Response.', [
+            'statusCode' => $response->getStatusCode(),
+            'reasonPhrase' => $response->getReasonPhrase(),
+            'headers ' => $response->getHeaders(),
+            'body' => $response->getBody()->getContents(),
+        ]);
+
+        // retry request if true
+        if ($this->checkResponse($response)) {
+            return $this->send($request, $options);
+        }
+
+        return $response;
+    }
+
+    /**
+     * Returns true whenever request should be retried.
+     *
+     * @param ResponseInterface $response
+     *
+     * @return bool
+     *
+     * @throws \Vault\Exceptions\ClientException
+     * @throws \Vault\Exceptions\ServerException
+     * @throws \RuntimeException
+     * @throws \Psr\Cache\InvalidArgumentException
+     */
+    protected function checkResponse(ResponseInterface $response)
+    {
+        // start re-authentication process if got 403 code and token is expired
+        if (
+            $response->getStatusCode() === 403 &&
+            !$this->isReAuthenticationInProgress() &&
+            $this->isTokenExpired($this->token) &&
+            $this->reAuthenticate()
+        ) {
+            return true;
+        }
+
+        if ($response->getStatusCode() >= 400) {
+            $message = sprintf(
+                "Something went wrong when calling Vault (%s - %s)\n%s.",
+                $response->getStatusCode(),
+                $response->getReasonPhrase(),
+                $response->getBody()->getContents()
+            );
+
+            if ($response->getStatusCode() >= 500) {
+                throw new ServerException($message, $response->getStatusCode(), $response);
+            }
+
+            throw new ClientException($message, $response->getStatusCode(), $response);
+        }
+
+        return false;
+    }
+
+    /**
+     * @return bool
+     */
+    private function isReAuthenticationInProgress()
+    {
+        return $this->reAuthenticationCounter > 0;
+    }
+
+    /**
+     * @param Token $token
+     *
+     * @return bool
+     */
+    private function isTokenExpired($token)
+    {
+        return !$token ||
+            (
+                $token->getCreationTtl() > 0 &&
+                time() > $token->getCreationTime() + $token->getCreationTtl()
+            );
+    }
+
+    /**
+     * @return bool
+     *
+     * @throws \Psr\Cache\InvalidArgumentException
+     */
+    protected function reAuthenticate()
+    {
+        while ($this->reAuthenticationCounter < $this->reAuthenticationThreshold) {
+            $this->reAuthenticationCounter++;
+
+            try {
+                $this->logger->debug('Trying to re-authenticate.');
+
+                if (!$this->authenticate()) {
+                    throw new \RuntimeException('Cannot re-authenticate.');
+                }
+
+                return true;
+            } catch (\Exception $e) {
+                // just skip all exceptions
```

Comment:
IMHO, it's important to save attempts count only in the case of failed authentication, so I think `$this->reAuthenticationCounter++;` should be here, not in general function.